typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1002 @@
+};
+
+f(collectionArray, 1, 2); // [ 1, 2 ]
+f(collectionSet, 2, 2); // Set(1) { 2 }
+```
+
+Of course this is a tiny set of what you can do with kinds.
+Most of all this part is inspired by [fp-ts](https://gcanti.github.io/fp-ts/) library and [Yuriy Bogomolov](https://ybogomolov.me/) blog.
+Fp-ts contains tone of things which you may know from Haskell and in blog you can find great explanations of it.
+Also, there are a lot of libraries based on fp-ts: [io-ts](https://github.com/gcanti/io-ts), [parser-ts](https://github.com/gcanti/parser-ts), [monocle-ts](https://github.com/gcanti/monocle-ts), [remote-data-ts](https://github.com/devex-web-frontend/remote-data-ts), etc.
+
+### Peano numbers
+
+Moving deeper we can think about type level programming and more concretely about computational on types.
+The basic primitive which may help us in this problem is Peano numbers.
+Lets how it may look in TypeScript.
+
+```typescript
+type Zero = "zero";
+
+type Nat = Zero | { n: Nat };
+```
+
+And now we are able to define `Succ` which add one to our `Nat`.
+```typescript
+type Succ<N extends Nat> = { n: N };
+```
+
+With simple `Nat` definition we are able to make smth useful.
+Let's try to create type safe vector which will store its length in type.
+
+Empty vector will be simply be a `"nil"`.
+
+```typescript
+type Nil = "nil";
+```
+
+Type `Cons` will add a value to the vector and additionally increase its type level length.
+
+```typescript
+type Cons<A, N extends Nat> = {a: A, v: Vec<A, N>};
+```
+
+And the main type `Vec` combine them with conditional type.
+Using word `infer` we are able to infer new type variable `R` from `N`.
+So, when `N` is a `Succ` we can infer nested `R` from it and return `Cons<A, R>`.
+Otherwise, it is simply `Nil`.
+
+```typescript
+type Vec<A, N extends Nat> = N extends Succ<infer R> ? Cons<A, R> : Nil;
+```
+
+And here are examples of its usage.
+We define two helper function which build `Vec` values.
+And as you can see may use it safely.
+
+```typescript
+const emptyVec: <A>() => Vec<A, Zero> = () => "nil";
+
+const pushVec: <A, N extends Nat>(a: A, v: Vec<A, N>) => Vec<A, Succ<N>> = (a, v) => {
+  return {a, v}
+};
+
+let empty: Vec<number, Zero> = emptyVec(); // Ok
+let oneElem: Vec<number, Succ<Zero>> = pushVec(1, empty); // Ok
+let twoElems: Vec<number, Succ<Succ<Zero>>> = pushVec(2, oneElem); // Ok
+let twoElemsInvalid: Vec<number, Succ<Zero>> = pushVec(2, oneElem); // error
+oneElem = twoElems; // error
+```
+
+This was a simple example what we can do with type level programming.
+Using such primitives we can implement different operations on types and even Fibonacci [implementation](https://mjj.io/2021/03/29/type-level-programming-in-typescript/) for example.
+Follow Mathias Jean Johansen [blog post](https://mjj.io/2021/03/29/type-level-programming-in-typescript/) or mentioned above Yuriy Bogomolov [examples](https://github.com/YBogomolov/talk-typelevel-ts/blob/master/src/typelevel.d.ts).
```

Comment:
Linking one blog post two lines in a row is a big no-no, can you squeeze these two lines into one?